File: swift/Sources/SignalClient/zkgroup/GroupSecretParams.swift
//
// Copyright 2020-2021 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

import Foundation
import SignalFfi

public class GroupSecretParams: ByteArray {

  public static func generate() throws -> GroupSecretParams {
    return try generate(randomness: Randomness.generate())
  }

  public static func generate(randomness: Randomness) throws -> GroupSecretParams {
    return try randomness.withUnsafePointerToBytes { randomness in
      try invokeFnReturningSerialized {
        signal_group_secret_params_generate_deterministic($0, randomness)
      }
    }
  }

  public static func deriveFromMasterKey(groupMasterKey: GroupMasterKey) throws -> GroupSecretParams {
    return try groupMasterKey.withUnsafePointerToSerialized { groupMasterKey in
      try invokeFnReturningSerialized {
        signal_group_secret_params_derive_from_master_key($0, groupMasterKey)
      }
    }
  }

  public required init(contents: [UInt8]) throws {
    try super.init(contents, checkValid: signal_group_secret_params_check_valid_contents)
  }

  public func getMasterKey() throws -> GroupMasterKey {
    return try withUnsafePointerToSerialized { contents in
      try invokeFnReturningSerialized {
        signal_group_secret_params_get_master_key($0, contents)
      }
    }
  }

  public func getPublicParams() throws -> GroupPublicParams {
    return try withUnsafePointerToSerialized { contents in
      try invokeFnReturningSerialized {
        signal_group_secret_params_get_public_params($0, contents)
      }
    }
  }

}

File: java/shared/java/org/signal/libsignal/zkgroup/groups/GroupSecretParams.java
//
// Copyright 2020-2021 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

package org.signal.libsignal.zkgroup.groups;

import static org.signal.libsignal.internal.FilterExceptions.filterExceptions;
import static org.signal.libsignal.zkgroup.internal.Constants.RANDOM_LENGTH;

import java.security.SecureRandom;
import org.signal.libsignal.internal.Native;
import org.signal.libsignal.zkgroup.InvalidInputException;
import org.signal.libsignal.zkgroup.internal.ByteArray;

public final class GroupSecretParams extends ByteArray {

  public static GroupSecretParams generate() {
    return generate(new SecureRandom());
  }

  public static GroupSecretParams generate(SecureRandom secureRandom) {
    byte[] random = new byte[RANDOM_LENGTH];
    secureRandom.nextBytes(random);

    byte[] newContents = Native.GroupSecretParams_GenerateDeterministic(random);

    try {
      return new GroupSecretParams(newContents);
    } catch (InvalidInputException e) {
      throw new AssertionError(e);
    }
  }

  public static GroupSecretParams deriveFromMasterKey(GroupMasterKey groupMasterKey) {
    byte[] newContents =
        Native.GroupSecretParams_DeriveFromMasterKey(groupMasterKey.getInternalContentsForJNI());

    try {
      return new GroupSecretParams(newContents);
    } catch (InvalidInputException e) {
      throw new AssertionError(e);
    }
  }

  public GroupSecretParams(byte[] contents) throws InvalidInputException {
    super(contents);
    filterExceptions(
        InvalidInputException.class, () -> Native.GroupSecretParams_CheckValidContents(contents));
  }

  public GroupMasterKey getMasterKey() {
    byte[] newContents = Native.GroupSecretParams_GetMasterKey(contents);

    try {
      return new GroupMasterKey(newContents);
    } catch (InvalidInputException e) {
      throw new AssertionError(e);
    }
  }

  public GroupPublicParams getPublicParams() {
    byte[] newContents = Native.GroupSecretParams_GetPublicParams(contents);

    try {
      return new GroupPublicParams(newContents);
    } catch (InvalidInputException e) {
      throw new AssertionError(e);
    }
  }
}
